content/blog/2022-03-22-type-alias/index.md

```diff
@@ -0,0 +1,210 @@
+
+
+## Stengths and Weaknesses of the paper
+This TBAA paper took a very concrete approach with examples from Modula-3 syntax, as well as three different performance analysis. One advantage of this is that the paper is fairly self-contained, and doesn't require highly specialized knowledge to understand. Thus the paper is more accessible to researchers from other areas, as well as people who want to implement TBAA but don't have a super strong techincal background. However, as a consequence, the paper doesn't explicitly provide a general theoretical framework that can be easily adopted by future programming languages and concepts.
+
+Limit analysis is a very neat idea, and it should definitely be incorporated into the empirical evaluation of other research projects. Having eight benchmark programs doesn't give the most convincing result, but the fact the programs are all real, practical ones does help.
+
+Finally, what the paper lacks the most is a concrete comparison between the effectiveness of this TBAA against other forms of alias analysis. Although the authors have touched upon how TBAA is usually more efficient, actual numbers and data points would be much more convincing.
+
+
+## TBAA in the Modern Computing Landscape
+
+The simplicity, speed, and accuracy of TBAA has allowed it to stick around in many modern compilers. LLVM has several different alias analyses, including an implementation of TBAA. The compiler frontend must include type annotations that can then be used for alias based optimization within LLVM itself. With a proper understanding of undefined behavior, TBAA can even be implemented for C and C++. Given the continued use of TBAA in real compilers, it is clear that TBAA works well in practice. TBAA can also be used in conjunction with flow sensitive alias analysis to improve accuracy in some cases.
+
+
+## Related Readings
+Markin and Ermolitsky<sup>[3]</sup> discussed their implementation of a simple type-based alias analysis, called strict aliasing, for a compiler from C/C++ to Elbrus, a general-purpose VLIW(very long instruction word) processor. Their experimental results show promising speedups in intraprocedural compilation and ok results for program-wide compilation. 
+
+Ireland, Amaral, Silvera, and Cui develeoped "SafeType"<sup>[4]</sup>, a tool to identify void pointer castings in C/C++. The TBAA paper mentioned that C/C++ language allows unsafe pointer casting and thus requires extra conservative alias analysis, but it didn't really address the issue. SafeType supposedly identifies ``violations on the C standard's restrictions on memory accesses'' with a flow-sensitive approach and enables TBAA on C/C++ programs that could previously lead to false results. The authors remarked that their purely static analysis doesn't introduce runtime overhead, which is problem of EffectiveSan<sup>[5]</sup> by Duck and Yap, which adds type safety through dynamically typing C/C++.
```

Comment:
I'm not 100% sure what you mean by "false results"… is this incorrect/unsound aliasing decisions?

Reply:
Yes, I think so